XR Support for TAA

robbiesri opened the pull request:
Since the engine now supports stereo jittered projection matrices, along with stereo motion vectors, TAA can now be enabled with XR.  This is a first attempt at adding XR Support for TAA to the Post Processing Stack v2.

The main changes live in:
* PostProcessLayer.cs
* TemporalAntialiasing.cs
* TemporalAntialiasing.shader

There are a few infrastructure changes:
* Make sure that the PostProcessRenderContext returns the correct width in case we are running a double-wide texture.
* Add OnPreRender in order to configure code/state correctly for the right eye.  This means some refactoring to share code between OnPreCull and OnPreRender.
* Shader changes to the TAA shader in order to properly sample screen-space textures.

There is work remaining to be done:
* Stereo instancing isn't supported here yet, as I need to add support for creating texture arrays as render textures.  I think I'd like to do this as part of a XR Instancing PR.
* I think a general XR fixup for the PostProcessing v2 stack would be useful.  It might have been worth doing 'before' this PR, but this acts as vertical slice.
* It's too bad we don't have an explicit multi vs single pass query on the scripting side.  I'm able to get by just fine w/o it, but it still feels unusual to divine the state instead of just having it exposed.

robbiesri committed:
Use non-deprecated EventTypes

2017.3 complains about some deprecated EventTypes used.  The 'correct' versions have long been available, so safe to switch over.

robbiesri committed:
Get context width/height from XR eye texture (when XR is active)

This handles the double-wide issue for temporary RTs in stereo.

We still need to accomodate:
* Stereo render target arraus
* Shaders that need to know the per-eye texture size

robbiesri committed:
Minor XR shader cleanup

Use TRANSFORM_TEX.
Also, there is the possibility that texcoord is being transformed in frag
too often.  However, it makes sense for tex-coord manipulation in frag.
Even so, I should check to make sure transforms are happening somewhere.

robbiesri committed:
WIP for TAA + stereo

* Shaders are mostly done (clamping needs to be amended, and maybe usage of _ST float4s)
* OnPreRender used for right eye in multi pass
* Setting up jittered matrices for stereo proj
* detecting stereo and enabling Single-pass path

What remains is configuring the history textures correctly for TAA.

robbiesri committed:
TAA works with Multi and Single Pass Stereo

Fair bit of hacking done, need to do a lot of cleanup.

RT arrays aren't addressed yet.

robbiesri committed:
Cleanups for TAA

Couple more bits to do

* Support RT arrays
* Generalize code between OnPreCull and OnPreRender for TAA

robbiesri committed:
Shader cleanups

Make sure coordinates are clamped where appropriate for TAA samples

robbiesri committed:
PostProcessLayer: Common code for OnPreCull and OnPreRender for Stereo

OnPreRender exists to hook into the right eye for multi-pass, and my initial implementation had mostly the same code as OnPreCul.  Now the two functions have a function to call for their common code, BuildCommandBuffers.

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_taa

robbiesri commented:
Ah, found the merge conflict.  Merged in origin/v2, fixed the conflict, pushed update to branch.

robbiesri committed:
Remove comment

numega commented on PostProcessing/Runtime/Effects/TemporalAntialiasing.cs:
```diff
@@ -96,18 +104,54 @@ public Matrix4x4 GetJitteredProjectionMatrix(Camera camera)
+                if (XR.XRSettings.isDeviceActive)
+                    rt.name = "Temporal Anti-aliasing History id #" + id.ToString() + " for eye " + context.xrActiveEye.ToString();
+                else
+                    rt.name = "Temporal Anti-aliasing History id #" + id.ToString();
```
Nit: Could we do this instead:
rt.name = "Temporal Anti-aliasing History id #" + id.ToString();
if (XR.XRSettings.isDeviceActive)
    rt.name += " for eye " + context.xrActiveEye.ToString();

robbiesri replied:
Yep, will fix this

numega commented on PostProcessing/Runtime/Effects/TemporalAntialiasing.cs:
```diff
@@ -116,15 +160,19 @@ RenderTexture CheckHistory(int id, PostProcessRenderContext context)
+                    rt2.name = "Temporal Anti-aliasing History id #" + id.ToString() + " for eye " + context.xrActiveEye.ToString();
+                else
+                    rt2.name = "Temporal Anti-aliasing History id #" + id.ToString();
+
```
We should create a new function since it's done here again.

numega commented on PostProcessing/Runtime/PostProcessRenderContext.cs:
```diff
@@ -7,7 +7,43 @@ public sealed class PostProcessRenderContext
+                    if (m_xrSinglePass && (xrDesc.dimension != TextureDimension.Tex2DArray))
+                        m_xrSingleEyeWidth = m_width / 2;
+                    else
+                        m_xrSingleEyeWidth = m_width;
```
Can we just use XRSettings.eyeTextureWidth? This value is only for one eye.

robbiesri replied:
Gah, you are right, thanks for catching this

numega commented on PostProcessing/Runtime/Utils/RuntimeUtilities.cs:
```diff
@@ -383,6 +386,75 @@ public static Matrix4x4 GetJitteredOrthographicProjectionMatrix(Camera camera, V
+            var bTan = (-1.0f + origProj[1, 2]) / origProj[1, 1];
+
+            float tanVertFov = Math.Abs(tTan) + Math.Abs(bTan);
+            float tanHorizFov = Math.Abs(lTan) + Math.Abs(rTan);
```
Why don't we just use origProj.decomposeProjection instead?

robbiesri replied:
I totally forgot we did that, thanks for the reminder.

robbiesri committed:
Remove more leftover comments

Chman commented:
I haven't had time to go deep into this VR yet but the first thing I noticed is that it doesn't have any compile guards. As far as I know, this will only work with 2017.3+, right? The stack needs to be compilable with 5.6+. If I were to merge this PR today it would break v2 for anyone not running 2017.3.

robbiesri commented:
@Chman 
Ah, you are totally right.  In my head, I was targeting 2017.2, but I forgot that the jittered stereo APIs are 2017.3 (which is embarassing).  I'll fix up the code with compile guards, thanks.

robbiesri committed:
Use decomposeProjection + cleanups

We recently added Matrix4x4.decomposeProjection, which gives us the frustum planes for the projection matrix.  I can then offset those with the jitter value, and create the jittered projection matrix.

Also a couple cleanups from PR feedback

robbiesri committed:
Gate XR TAA for 2017.3 and newer

Camera.CopyStereoDeviceProjectionMatrixToNonJittered and Camera.GetStereoNonJitteredProjectionMatrix are only on 2017.3, so we can't really support XR TAA before that.

Also, fixed a method name typo

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_taa

robbiesri committed:
Amend command buffer clear comment

For XR, it might be worth investigating retaining command buffer contents for multi-pass right eye.  But probably not, it's the 'old' technique.

robbiesri commented:
Ok, I took in the feedback from reviewers.

I actually have a lot for architectural change stuff coming up in my general XR change.  There are a couple things I'd like to change here that I'll defer to the broader change (using RenderTextureDescriptor, getting rid of singleEyeWidth, more stuff).

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_taa

robbiesri committed:
Use new texcoordStereo in TAA shader

robbiesri commented:
I merged origin/v2 again, mostly because of the TAA infrastructure changes, and to latch onto the new 'texcoordStereo' parameter.  Woo!

Chman committed:
Fixed compatibility with 5.6 and 2017.1

Chman commented:
In its current state this PR is good but doesn't compile on anything lower than 2017.2. I pushed a few compatibility fixes & tweaks in #301, I'll deal with conflicts after this is fixed.

robbiesri committed:
Merge pull request #301 from Unity-Technologies/xr_taa_compatibility

XR Support for TAA (fixes)

robbiesri committed:
Extra cleanups after xr_taa_compatibility

robbiesri committed:
Merge remote-tracking branch 'origin/v2' into xr_taa
